cluster-autoscaler/cloudprovider/aws/aws_manager.go

```diff
@@ -80,41 +87,179 @@ func createAWSManagerInternal(configReader io.Reader, service *autoScalingWrappe
 		}
 	}
 
+	asgs, err := newAutoScalingGroups(*service)
+	if err != nil {
+		return nil, err
+	}
+
+	cfgs, err := discoveryOpts.ParseASGAutoDiscoverySpecs()
+	if err != nil {
+		return nil, err
+	}
+
 	manager := &AwsManager{
-		asgs:      newAutoScalingGroups(*service),
-		service:   *service,
-		interrupt: make(chan struct{}),
+		service:         *service,
+		asgs:            asgs,
+		autoASGs:        cfgs,
+		neverUnregister: make(map[AwsRef]bool),
 	}
 
-	go wait.Until(func() {
-		manager.asgs.cacheMutex.Lock()
-		defer manager.asgs.cacheMutex.Unlock()
-		if err := manager.asgs.regenerateCache(); err != nil {
-			glog.Errorf("Error while regenerating Asg cache: %v", err)
-		}
-	}, time.Hour, manager.interrupt)
+	if err := manager.fetchExplicitAsgs(discoveryOpts.NodeGroupSpecs); err != nil {
```

Comment:
For me, it wasn't very straight-forward to understand why we have to pass auto discovery specs and "explicit" discovery specs separately like this.
Can we pass the `discoverySpecs.NodeGroupSpecs` while creating the `AwsManager` struct above, so that it looks more consistent between auto vs explicit discovery specs in regard to those two are both about discovery?

Reply:
We certainly could store `discoverySpecs.NodeGroupSpecs` in the `AwsManager`, but it would only be for consistency's sake because we need fetch 'explicit' node groups only once - at the creation of the AwsManager.

I think you're suggesting an alternative like this:
```go
func createAWSManagerInternal(...) (*AwsManager, error) {
	manager := &AwsManager{
		service:         *service,
		asgs:            asgs,
		autoASGs:        cfgs,
                explicitASGs:    discoveryOpts.NodeGroupSpecs,
		neverUnregister: make(map[AwsRef]bool),
	}
        manager.fetchExplicitAsgs()
}
```

To me this alternative makes it less obvious that `discoveryOpts.NodeGroupSpecs` is only consumed once, by `fetchExplicitAsgs().

In this PR we have three 'flavours' of spec:
1. A regular `--nodes' 'min:max:name` style spec. `NodeGroupSpecs` is a `[]string` slice of these specs. Each of these specs is passed to `dynamic.SpecFromString` when `fetchExplicitAsgs` is called.
2. A `--node-group-autodiscovery` ASG spec, like `asg:tag=coolTag`. This is parsed into an `[]ASGAutoDiscoveryConfig` slice by `discoveryOpts.ParseASGAutoDiscoverySpecs()`.
3. A `-node-group-autodiscovery` MIG spec like `mix:prefix=coolPrefix,minNodes=0,maxNodes=100`. This is parsed into a `[]MigAutoDiscoveryConfig` slice by `discoveryOpts.ParseASGAutoDiscoverySpecs()`.

These three can never be completely consistent, because they do slightly different things:
1. Contains everything we need to know to register in ASG/MIG, and is only consumed once.
2. and 3. instead contain inputs to discover ASGs/MIGs to register, and are consumed repeatedly over the life of the CA.

Talking through this, I wonder whether `dynamic.NodeGroupSpec` should be merged with `NodeGroupDiscoveryOptions`. This would move all of the flag/'spec' parsing code up to `NodeGroupDiscoveryOptions`.

This would look something like:
```go
func (o NodeGroupDiscoveryOptions) ParseExplicitSpecs() ([]NodeGroupSpec, error) {}

func (m *AwsManager) fetchExplicitAsgs(specs cloudprovider.NodeGroupSpec[]) error {}
```

What do you think? Or is there some other approach we could take to make this clearer?